### ISSUE TITLE
Request: KopiaUI / Kopia Server uses low value for parallelisation

### ISSUE THREAD

**Issue by LordMike** [kopia/kopia#1151]
Hi,

I'm really impressed with this project. I'm running it across three devices to keep backups and have a restore capability outside of the device that will have crashed :).

When performing a large-ish backup with B2 (and Wasabi), I noticed that my upload bandwidth is jumping a lot, indicating to me that there is a time period inside the application where something is being prepared. The upload isn't one continuous operation. When I inspect the logs, I can also see gaps in the files it operates on, like below. These gaps of time, along with the sporadic upload speeds lead me to thinking of pipelining.

If this is already implemented, I'm not sure if B2 / Wasabi is _just that slow_ ... :P

```
17:27:29.054	could not find cache entry for 20180627_100225601_iOS.jpg
... 42 files within 3 seconds ...
17:30:50.905	could not find cache entry for 20180701_143953919_iOS.jpg

.. gap of 1-6 seconds with no logs ..

17:31:44.075	could not find cache entry for 20180701_203002000_iOS.MOV
17:34:05.309	could not find cache entry for 20180701_203625000_iOS.MOV
17:34:57.400	could not find cache entry for 20180701_203943000_iOS.MOV
17:35:52.344	could not find cache entry for 20180701_204549000_iOS.MOV
17:36:09.261	could not find cache entry for 20180701_211330784_iOS.jpg
.. 90 lines within ~6 seconds ..
17:42:10.125	could not find cache entry for 20180709_183427000_iOS.MOV

.. gap of 1-4 seconds with no logs ..

17:45:03.109	could not find cache entry for 20180711_075152263_iOS.jpg

.. etc.
```

One way to pipeline the uploads is to queue packs to upload (keep a limited number in memory to reduce usage), to ensure that the internet connection is constantly used. There will be some extra logic and state tracking to keep track of the packs to upload and when they're done - and to create packs as needed.. And I won't event begin to speculate on the logic needed to create packs ahead of time, for files as they're being deduplicated :)

Mike.

**Comment by ChetanGoti** [comment]
I'm syncing to S3 with `--parallel=5` and the bandwidth utilization is constant throughout. 

Did you try with increasing `--parallel` flag which defaults to `1`?

Reference: https://kopia.io/docs/reference/command-line/common/repository-sync-to-b2/

**Comment by LordMike** [comment]
Omg... it defaults to 1.. :|

Lemme do some tests.

**Comment by LordMike** [comment]
So for some numbers. I've created a new B2 bucket, with no serverside encryption. I've also created a new repo for the new bucket, without compression and using the FIXED-4M splitter (to reduce the effect of the other one I'm using).

Using the b2 CLI, I can consistenly reach approximately ~250 Mbps with varying levels of parallelism and part sizes. If I use the Kopia UI, I can seemingly only hit 25-70 Mbps, in spikes. I've downloaded the Kopia CLI for Windows, and using this, I was actually able to hit 250-500 Mbps at time, using a parallelism factor of 20 (just for testing). At no time is Kopia CPU or disk -bound, it's purely networking.

I'll check if it's possible to tweak the parallelism of Kopia UI.

But from here, it seems that Kopia itself isn't the issue, or at least it has workarounds.

Thanks. :)

**Comment by LordMike** [comment]
I've looked into the Kopia server component (as the electron app just wraps that). Starting the server manually, using f.ex. `.\resources\server\kopia.exe server --insecure --server-password=kopia --log-level debug` shows a log output that indicates a pararalelism of 1 is actually being used. At least, there is a `PutBlob` log line every few seconds, but its just that. Theres no PutBlob's in between the time it takes for one blob to upload and the next.

I'm not entirely proficient in Go, so I can't really follow the code. I've made it from [`handleSourcesCreate`](https://github.com/kopia/kopia/blob/4b251bdaaccff70c0f1b85e825bd56a79c6b0219/internal/server/api_sources.go#L44) which I think receives the API call to create a new snapshot, and down to [source_manager's snapshotInternal](https://github.com/kopia/kopia/blob/4b251bdaaccff70c0f1b85e825bd56a79c6b0219/internal/server/source_manager.go#L239-L312) which I think does the acutal uploading. I'm not sure where the `NewUploader` (`u`) is configured, if at all - if the configuration is not done at all, that could explain why the parallelism is set so low.
